### ISSUE TITLE
Future of Mono

### ISSUE THREAD

**Issue by Maniues** [mono/mono#20931]
Hello
I would like to ask about the future of mono.
Specifically - the .NET Framework is no longer being developed. The Mono page describes "Open Source .NET Framework". Will mono die with the .NET Framework then and your only options will be .NET 5 and .NET 6 or will you continue to develop it?

**Comment by marek-safar** [comment]
The Mono project originally focused on .NET Framework compatibility and it's something that I don't see changing in the future.

For Mono plans inside .NET Core setup please read https://devblogs.microsoft.com/dotnet/introducing-net-5/ which explains well how MonoVM will work inside .NET 5+.

**Comment by Maniues** [comment]
So mono will be still developed independently of .net 5.0 and above?

**Comment by marek-safar** [comment]
Yes, the packages coming from mono/mono will be independent of .NET5+ and .NET5+ MonoVM will live in dotnet/runtime.

**Comment by Maniues** [comment]
I meant more about whether at some point there will be a decision to eliminate mono fragments in order to make .NET (dotnet / runtime) a less active and complete implementation of .NET. I know mono is one of the .NET dotnet / runtime dependencies so it is unlikely to go away but may become more as a dependency / library rather than a full user friendly implementation

**Comment by marek-safar** [comment]
Please read the blog it has all the details. There is no plan right now to eliminate MonoVM in .NET5+

**Comment by Maniues** [comment]
Hi
I am not talking about eliminating mono in .NET 5.0 and above, but about eliminating mono. Mono is a great implementation of .NET, but MS is striving to unify .NET, which is why .NET Framework and .NET Core have died (yes, I know with Core only the name has changed). If Microsoft wants to make .NET transparent and uniform (as it used to be, in the days of only the .NET Framework) then it may turn out that mono will be limited to certain elements necessary for the operation of dotnet / runtime, sdk and AOT. Anyway, probably for the sake of unifying one IDE for .NET has already been killed on Windows and MacOS.
But assuming there are no plans to reduce the mono scale, how are you going to work on mono? It was an implementation of the .NET Framework that is no longer being developed. Will you be introducing new features or will you rather fix bugs and improve stability?

**Comment by marek-safar** [comment]
The Mono Project is an open-source project and anyway can contribute bugs or new features (for example M1 support landed just recently). As I explained earlier Microsoft's focus is on .NET5+ version but it will still support existing products and customers using mono/mono codebase.

**Comment by Adahel** [comment]
> The Mono Project is an open-source project and anyway can contribute bugs or new features (for example M1 support landed just recently). As I explained earlier Microsoft's focus is on .NET5+ version but it will still support existing products and customers using mono/mono codebase.

@marek-safar Microsoft will be right not to invest in Mono in the future, as it is an implementation of an obsolete .Net Framework technology. Soon, there may be no team to accept pull requests for the Mono project, and without a Mono fork, the project will die. This question came at a great time, as Microsoft has been selective in which pull requests to accept, although there is a demand for Windows Forms, Windows Forms pull requests are being completely ignored. I don't know if the pull requests were ignored because of the development of .Net 6 or an internal policy, I won't speculate on the ignored pull requests. I have just a few suggestions: The developers who collaborate on this project are human beings who invest their time, they do not deserve silence. The silence keeps the developers away, my suggestion is to make it clear what will or will not be accepted. When a project is no longer financially profitable, donate the project to an open source organization like Eclipse or Apache, as I believe that only the community should decide when an open source project dies. If there is no donation, on his death, archiving the repository so as not to give false hopes. **This comment refers to the Mono .Net Framework and not .Net 5 or .Net 6.**

**Comment by Maniues** [comment]
@Adahel Thanks for writing this, because you made me realize that my fears are right. It's true that MS doesn't invest in Mono while keeping it alive. Unfortunately, the truth is that Mono will die not even from the fact that it comes under Microsoft, but because it was created as an implementation of the .NET Framework. It is just like Wine. If windows died, what would wine be? What would it be for? They could develop compatibility, but over time they will be at the point where Mono is now. Unfortunately, Mono has little functionality here, as it's possible that developers would like to add new features, but unfortunately compatibility with the .NET Framework is more important. I ignore the fact of dependence on MS, because probably everyone has found out on the example of M#noD#velop 

>When a project is no longer financially profitable, donate the project to an open source organization like Eclipse or Apache, as I believe that only the community should decide when an open source project dies. If there is no donation, on his death, archiving the repository so as not to give false hopes. **This comment refers to the Mono .Net Framework and not .Net 5 or .Net 6.**

Yes, I agree, but there is one fact ... Mono is in this community right now and is unlikely to release him. Ahhh, if only the .NET Foundation was independent...

**Comment by madewokherd** [comment]
Funny you should mention Wine, we still have to support .NET Framework applications, and we use (a fork of) Mono to do that. But for .NET 5 it seems like the best approach for us is to use .NET 5.

If there's still a demand for winforms on other platforms, maybe it'd make sense to pull that out of Mono so it can be used with newer .NET?

**Comment by Maniues** [comment]
> Funny you should mention Wine, we still have to support .NET Framework applications, and we use (a fork of) Mono to do that. But for .NET 5 it seems like the best approach for us is to use .NET 5.


Why it's funny?

**Comment by madewokherd** [comment]
It sounds like it's one of the few remaining use cases for Mono as a .NET Framework implementation specifically, where you wouldn't ideally want a newer version of .NET.

**Comment by Adahel** [comment]
> If there's still a demand for winforms on other platforms, maybe it'd make sense to pull that out of Mono so it can be used with newer .NET?

A multiplatform implementation of Windows Forms for modern .Net does not seem to make sense with the arrival of .Net Maui. The question is about ignored pull requests and the end of Mono's life. **When I speak Mono, I am referring to the implementation of .Net Framework 4.8 and below.**

**Comment by marek-safar** [comment]
> Windows Forms pull requests are being completely ignored.

@Adahel could you point me to specific PRs you would like to merge. The Mono team does not have the capacity to test and fully review such PRs so one alternative could be to automatically merge WinForms PRs.

**Comment by Maniues** [comment]
@Adahel  I have looked a little deeper at your statement. It turns out that .NET 5/6 uses Mono to extend cross-platform to mobile platforms (e.g. iPhone, Android). MS is unlikely to kill mono completely (at least for now), because it plays a role that cannot be ignored at the moment (thousands of Unity games for android / ios). As I said before - if you ONLY implement a platform, and you do not set directions for it, if it dies, the development of your implementation will stop at some point. Currently what I see for mono is some continuation of the .NET Framework, .NET Core / 5/6 compatibility libraries, purely technical bug fixes, adding support for new platforms and newer versions of C #. At some point, either mono will die or be developed by other organizations. It is also possible that Microsoft will set a new direction for him, but I sincerely doubt it. Currently, only Mono stands in the way of MS unification as a viable alternative platform.
@marek-safar How can we be sure (we, the C # applications developers), that Mono will be supported and MS will not kill the project, or reduce it to the platform compatibility layer scale?

**Comment by Adahel** [comment]
@marek-safar

> so one alternative could be to automatically merge WinForms PRs.

I believe that the tests can be postponed, but a minimum review is always necessary.



> @Adahel could you point me to specific PRs you would like to merge.

I would indicate: [@sancheolz](https://github.com/mono/mono/pulls/sancheolz) and #20687 (@sancheolz, @NVoronchev)



> @marek-safar How can we be sure (we, the C # applications developers), that Mono will be supported and MS will not kill the project, or reduce it to the platform compatibility layer scale? What is our guarantee

@Maniues Companies are like countries that have their own interests, moments of instability and the interests of a company can change depending on the leadership. No company can guarantee anything, times change and leadership changes too.





> The Mono team does not have the capacity to test and fully review such PRs

With the development of .Net 6, this statement is understandable and with the comments on this issue, I believe that for now Mono will still have life, indirectly my question has been resolved. This issue may become relevant to me again after the release of .Net 6.

**Comment by Maniues** [comment]
@Adahel  I know, no one can guarantee anything to anyone. But why are we sure and "guaranteed" that .NET 5 and newer will live long and will evolve? And why do we think mono can die? This is what it is all about. If you are clearly investing in this technology and seeing the results of it then I'm sure an investment in .NET 5/6 / etc is a good idea. As for mono, MS gives a sign that it is better to depart from it (killing monod3v on many platforms), or to use it where necessary (multi-platform applications built with the .NET Framework). Today we know that the .NET Framework has succeeded. Will this be the case with .NET Core (5 and above)? Even if not, MS will do whatever it takes to push it through, and in the worst case it will fall back to Framework, but then only mono will be used.

**Comment by grendello** [comment]
@Maniues you (and others) raise a lot of valid points and worries about the future of Mono. But, the way I see it, the questions for guarantees that Microsoft will keep Mono alive and worries that it will "kill" the project are the sign of why Mono isn't thriving. Mono has always been an open-source project, licensed under a very liberal license and, as such, it cannot be "killed" - at any given time it can be forked and development can continue **if** (and that's the key here) there is a community behind it - community not of only users, but also active developers contributing to it but also looking at PRs, proposing new directions, adding features, e.g. compatibility with future .NET versions - there's no reason why mono/mono VM cannot be made to work with them other than the lack of people who can devote time to it. This is even easier given that dotnet/runtime has a Mono VM and it's not impossible to port changes made to that version of mono to the runtime in mono/mono. WinForms (and other) PRs were mentioned in the thread above with a question why they aren't merged, @marek-safar explained that we're simply too few to handle the workload that .NET6 development involves. However, this doesn't prevent a group of community contributors from creating an organization on GitHub and forking mono/mono (and potentially other repos from the Mono org) in there and accepting the aforementioned PRs there - not to divert from Mono, but to keep it alive and in sync with the current version in mono/mono - making it possible for mono/mono to eventually sync up to the forked version when there's enough time on "our" side to do the work. Mono used to have a very strong community of contributors (not only developers) which was passionate, dedicated and full of enthusiasm. That community unfortunately, no longer exists in the form it used to - reasons for this are history and this is not the place and time to review them (I'm sure you can DuckDuckGo it) and, thus, the responsibility of not only further development of the Mono VM (in dotnet/runtime and mono/mono) but also of handling PRs, issues etc rests on shoulders of a small team which can only do so much.

This is very similar to the situation with, e.g., zlib which was forked into zlib-ng for similar reasons and the two projects coexist. The same happened with MonoDevelop - it got forked to https://github.com/dotdevelop/dotdevelop because someone out there had the will, energy and drive to do it - to keep up what was left behind (for whatever reasons, they don't matter in this context). Recently, I took up a work to take over `Mono.Posix` maintenance and development by splitting it out of the mono/mono tree and putting it in its own repository (https://github.com/mono/mono.posix/) - since it's not part of dotnetcore but is used by people all over the world, including my team at Microsoft (Xamarin.Android). This is NOT something that my work contract requires me to do, we could get by by using the version from mono/mono but I care about mono/mono enough to do this work, while making it useful for my team. 

Please, do not read the words above as a criticism but rather as encouragement - to take up the slack and work with anyone interested in keeping mono/mono alive (including, I'm sure, many members of the Microsoft Mono team, past and present). I personally know I would love to see the Mono community back to its former glory (which spawned great projects like Banshee, F-Spot, Smuxi, MonoDevelop, Tomboy) and I would be happy to be part of it.

**Comment by grendello** [comment]
Regarding the plans, I can't answer that - I don't work on the Mono runtime team. And, let me point out - you once again ask "what will YOU" introduce in Mono. Invert that and ask yourself "what the community can introduce to Mono"? What YOU personally can/want/have time to do? Don't wait passively on "us" (the MS Mono etc teams), be proactive, help rebuild the community, revive projects that matter etc etc.

**Comment by grendello** [comment]
@Maniues many years ago I started to contribute to Mono to scratch my own itch - I wanted to have ASP.NET 2.0 on Linux. These were the golden years of Mono community, a LOT was happening - because people were active, were working with Mono The Project and not waiting on it to implement something.

**Comment by Maniues** [comment]
@grendello Great words.
**For me**, the topic is exhausted.
For a few days, maybe a week, maybe two, I will keep this topic active if someone wants to comment again, and because some other threads that may need to be answered.
If the administrators do not close this issue, it will be active as a gate to the continuation of the next thread (e.g. isolating the winforms and adapting it to the new .NET).
**I recommend creating a new thread or discussion and linking it here**, instead of continuing the thread here as it will become a mess.
Thanks to everyone involved

**Comment by michaldobrodenka** [comment]
Since NET5/6 platform support is limited, mono for me is the only way to have c# code running on ARMv6. (Now when mono is in undead zombie state, I have to deal with bugs and decide if I 
1. upgrade hw to ARMv7 - very expensive
2. wait for NET6+ to arrive to ARMv6 - is it real? 
3. Or try to fix/apply some "do not merge" PR, compile it, test it)

**Comment by michaldobrodenka** [comment]
> The Mono project originally focused on .NET Framework compatibility and it's something that I don't see changing in the future.
> 
> For Mono plans inside .NET Core setup please read https://devblogs.microsoft.com/dotnet/introducing-net-5/ which explains well how MonoVM will work inside .NET 5+.

Whan I saw thig blog, I understood it that we will get quality of MS NET5+ libraries with platform versatility of MonoVM - so basically NET5+ to mono platforms which seemed fantastic.

**Comment by marek-safar** [comment]
> I understood it that we will get quality of MS NET5+ libraries with platform versatility of MonoVM

That's still the direction we would like to go but it won't happen overnight. For .NET6 we are primarily focusing on mobile and helping the community with s390x MonoVM support. If you feel strongly about any other arch or platform you can contribute to help with the effort.

**Comment by michaldobrodenka** [comment]
@marek-safar Thank you for clarification! I don't feel strong in ARM arch/assembly. I've contributed only very slightly to mono source code years ago. I would be very happy to test/report bugs.

**Comment by ArsenShnurkov** [comment]
When csc compiler from mono will start to support C# 9 ?

https://github.com/mono/mono/issues/20830

https://github.com/dotnet/msbuild/issues/6058

https://stackoverflow.com/questions/65309660/csc-on-macos-does-not-list-c-sharp-9

https://developercommunity.visualstudio.com/t/mono-fails-to-build-multi-targeting-c-9-library/1308259

**Comment by Alovchin91** [comment]
We're migrating our .NET Framework codebase to .NET 6, but there are still scenarios that can only run on Mono. I wonder if there's any chance of a new Mono release that would be able to run .NET 6 assemblies? That would include C# 9 and 10 support.

**Comment by michaldobrodenka** [comment]
I made a proposal 
https://github.com/dotnet/sdk/issues/22560

> We're migrating our .NET Framework codebase to .NET 6, but there are still scenarios that can only run on Mono. I wonder if there's any chance of a new Mono release that would be able to run .NET 6 assemblies? That would include C# 9 and 10 support.

**Comment by michaldobrodenka** [comment]
> 
> 
> We're migrating our .NET Framework codebase to .NET 6, but there are still scenarios that can only run on Mono. I wonder if there's any chance of a new Mono release that would be able to run .NET 6 assemblies? That would include C# 9 and 10 support.

What are scenarios that run only in mono? Except platforms not supported by dotnet?

**Comment by ArsenShnurkov** [comment]
> What are scenarios that run only in mono?
Remoting and Managed Add in Framework (MAF)

**Comment by sancheolz** [comment]
> What are scenarios that run only in mono? 
WinForms for Linux, Remoting

**Comment by ericoporto** [comment]
I also need it for WinForms on Linux

**Comment by DeafMan1983** [comment]
@sancheolz If Gtk4Sharp or XSharp comes for Net 6.x under Linux. WinForms isn't good for Linux. And nice idea "MonoVM" like JVM, is it correctly?

Enjoy your development!

**Comment by lucapivato** [comment]
Aren't .NET 6 and 7 for iOS and Android still running on mono?

**Comment by srxqds** [comment]
> Aren't .NET 6 and 7 for iOS and Android still running on mono?

they are still running on mono

**Comment by DeafMan1983** [comment]
Yes but Microsoft will replace Mono with Dotnet version. Since I have tried Mono Runtime on Ubuntu 22.04. It looks weird and because I open Godot with Mono then it happens like Mono is dead now and Dotnet Webassembly doesn't use Mono Runtime now and Dotnet Webassembly has own Dotnet version.  Please stop and please take rest with Mono. We need rest Mono at cemetery.  Hello Dotnet!

**Comment by elhoir** [comment]
is dotnet open source?

**Comment by sancheolz** [comment]
> @sancheolz If Gtk4Sharp or XSharp comes for Net 6.x under Linux. WinForms isn't good for Linux. And nice idea "MonoVM" like JVM, is it correctly?
> 
> Enjoy your development!

@DeafMan1983 Switching to Gtk 4 Sharp or XSharp involves rewriting most of the code. For Remoting replace I found the project [CoreRemoting](https://github.com/theRainbird/CoreRemoting). And a [Core.System.Windows.Forms](https://github.com/DanielVanNoord/System.Windows.Forms) project for using Winforms from mono in a .net core environment. This is a temporary solution to leave for something else

**Comment by nextfool** [comment]
About 5 years ago, we upgraded most of our projects to .net core (and latter .net 5,6,7 etc) and switched our production servers from windows to linux. There is a legacy asp.net project(since .net mvc 3) however, is too critical and infeasible to update it to .net 6/7.  Since then mono was the only option for us.  Now we dev with windows 11 + vs 2022, and run on ubuntu + mono. As far as we know, .net 4.8 will be available on windows for a significant while, so we just hope that mono can live as long as .net 4.8.

**Comment by mirh** [comment]
So, I see that a [very short](https://github.com/mono/mono/graphs/contributors) while after the opening of this thread, activity here really took a nosedive (winding down of Xamarin I presume?). Yet some sort of development continued over at [dotnet/runtime](https://github.com/dotnet/runtime/tree/main/src/mono). 

And there might still remotely be some overlapping code.. but its purposes aren't really much compatible with the old version, isn't it? For as much as I [understand](https://learn.microsoft.com/en-us/dotnet/maui/what-is-maui#how-net-maui-works) that the runtime is still very important, there's none of the old class libraries (Windows.Forms among others) required to run 2.0 or 4.8. 
Shouldn't this be acknowledged somewhere? Like, it's not even a biggie if you don't have time anymore to review and babysit this repo for aged software (especially since it's not bringing the bread home) but you could at least issue a call for new maintainers. 
Both Esme and @joncham &co (I don't think @Unity-Technologies would tag anyone?) could probably be interested.

**Comment by madewokherd** [comment]
> Both Esme and @joncham &co (I don't think @Unity-Technologies would tag anyone?) could probably be interested.

I *might* be interested in reviewing patches from the community, but I don't want to end up in a position of responsibility with no power, which would be likely to happen in the unlikely event I were allowed to do that here. I'm also already maintaining a fork, and I don't want to be working in two divergent branches of the same project (although there are also some changes in my fork that specifically support embedding within Wine, which wouldn't be suitable in other contexts).

**Comment by mirh** [comment]
Is this specific choice of affairs just dictated by the lack of control over the platform, or did you actually get some feedback and opinion from the supposed maintainers of @mono?
Like, I don't know, not that you could or should be expected to go through 2K issues but it seems a pity to abandon this - let alone with absolutely 0 information for the general public about the actual status of the project.

**Comment by madewokherd** [comment]
I haven't contacted current mono maintainers privately (I don't even know whom I'd contact), I'm just assuming I wouldn't be able to work anything out with them. I'd pretty much be asking for complete trust to determine the future direction of the project (and I'm sure I'd make decisions they wouldn't like), from a group of people who barely engaged with me even when the project was active. And decision number one would still be "get everything off of their hosted infrastructure so the project can continue if jenkins/the azure ci/this github organization go away".

**Comment by madewokherd** [comment]
I guess a major factor is also that Winehq would prefer to have all their projects in their self-hosted Gitlab, and that includes Wine Mono, so I planned to do that move regardless. And at that point I really don't want to be responsible for two entirely separate sets of repos with all their own infrastructure, even if I could be sure that nothing's going to happen to the Mono organization.

**Comment by mirh** [comment]
That makes sense I guess (even though *wine-winemono* vs *linux-mono* or *whateverelse-mono* must not exactly be the smoothest shareable environments on the shed). 
With this said, I'll now just rudely summon everybody and their aunt because this silence is also kinda lame @akoeplinger @kg @lambdageek @vargaz @directhex @steveisok @marek-safar @BrzVlad @scott-ferguson-unity @bholmes

**Comment by madewokherd** [comment]
FWIW I think the vast majority of changes in Wine Mono's branch would be good for general use (or could have been if I'd had a reason to properly consider it). The actual diff is large right now, but it could be small in the future. My expectation when this is set up is that Wine developers should target the main branch for pretty much everything, not the wine-mono branch.

**Comment by sancheolz** [comment]
@madewokherd, I have some PRs in this repo https://github.com/mono/mono/pulls/sancheolz and some commits here https://github.com/DanielVanNoord/System.Windows.Forms, do you need them as MRs to yours wine-mono project?

**Comment by madewokherd** [comment]
Yes, please make them to the main branch here https://gitlab.winehq.org/wine-mono/mono. Thanks.

**Comment by madewokherd** [comment]
In case it interests you, there was a bugzilla product for mono and a component for winforms added yesterday: https://bugs.winehq.org/describecomponents.cgi?product=Mono

**Comment by madewokherd** [comment]
The Winehq fork is now the official upstream: https://github.com/mono/mono/issues/21796

I am going to have to request that issues and code contributions be refiled at Winehq for those who still care about them. I have enough on my plate just getting things up and running in their new home, along with my other duties at CodeWeavers, and I don't have the bandwidth to sort through that whole backlog.

**Comment by Neustradamus** [comment]
Dear @mono team, @jeffschwMSFT, etc.,

It is possible to add ownership of the @mono organization to Wine HQ team? -> @julliard
It is important to have a continuous update of the code on GitHub, a mirror of https://gitlab.winehq.org/mono.

It is important to have official mirrors like the main Wine sourcecode:
- https://github.com/wine-mirror/wine (mirror of https://gitlab.winehq.org/wine/wine)

Can you give all informations of hosting servers, domain names to the new team for the continuous service?
The new team must have all to the future.

Thanks in advance.

Regards.

Initally published here:
- https://github.com/mono/mono/issues/21796#issuecomment-2335189368
